### PR TITLE
Feature/term frequency fixes

### PR DESCRIPTION
**Pull request by BeritJanssen**
Instead of fitting a curve over the matches, this branch now goes back to sampling a sizeable batch of the data first, and estimating then by assuming linear decay between the end of the sampling interval and match 1. This means we'll probably overestimate (instead of underestimate) the real number of matches in some cases, but fitting an exponential curve over the data proved not feasible due to fluctuations in the data (see #959). We may still decide that the estimate of decay is not worth it and roll back the changes related so our estimate is back to `sampled_matches + n_not_analyzed_docs`. What do you think, @lukavdplas?

This branch does introduce one essential (frontend) fix though: sorting by relevance rather than date (or other sort, which used to be kept in the query model) when requesting the top hits in the `visualization.term_frequency.get_match_count` function.

### CODE REVIEW

File: frontend/src/app/visualization/barchart/timeline.component.html [hunk 1]
```diff
@@ -31,8 +31,9 @@
     <div class="message-body">
         <div class="block">
         The results for your search include too many documents to count all matches.
-        Only the first {{percentageDocumentsSearched}}% of documents for each bin are analyzed;
-        the rest are assumed to have only one match for your query.
+        Only a sample of the first {{percentageDocumentsSearched}}% of documents for each bin are analyzed;
+        the rest are assumed to contain linearly decreasing matches between the end of the sample
+        to one match.
```
Comment (lukavdplas) [inline frontend/src/app/visualization/barchart/timeline.component.html:36]: Perhaps just change this into "the match count for other documents is extrapolated" and link to the manual for a full description.

File: frontend/src/app/visualization/barchart/barchart.directive.ts [hunk 2]
```diff
@@ -582,6 +582,13 @@ export abstract class BarchartDirective
         return queryModelCopy;
     }
 
+    /** return a copy of a query model with removed sort parameter (=> relevance sorting) */
+    removeSort(query: QueryModel): QueryModel {
+        const queryModelCopy = query.clone();
+        queryModelCopy.sort.setSortBy(undefined);
```
Comment (lukavdplas) [inline frontend/src/app/visualization/barchart/barchart.directive.ts:588]:
```suggestion
        queryModelCopy.sort.reset();
```

File: backend/visualization/term_frequency.py [hunk 3]
```diff
@@ -61,22 +64,31 @@ def extract_data_for_term_frequency(corpus, es_query):
...
     skipped_docs = total_results - len(found_hits)
-    match_count = matches + skipped_docs
-
+    if not skipped_docs:
+        return n_matches
+    
+    mean_last_matches = sum(matches[-ESTIMATE_WINDOW:]) / ESTIMATE_WINDOW
+    # we estimate that skipped contain matches linearly decrease 
+    # from average in ESTIMATE_WINDOW to 1
+    estimate_skipped = int((mean_last_matches - 1) * skipped_docs / 2)
```
Comment (lukavdplas) [inline backend/visualization/term_frequency.py:90]: Isn't this:

```suggestion
    estimate_skipped = skipped_docs + int((mean_last_matches - 1) * skipped_docs / 2)
```

**Review by lukavdplas — COMMENTED** [review]
The code looks good, though I would think the formula for estimating the surface under the curve is different (see my suggestion).

I would think that `auc` would be quite fast since it also uses linear extrapolation, but I trust your judgement on this. The changes proposed here are quite modest so if it improves the accuracy, it's definitely worth it.

File: backend/visualization/term_frequency.py (hunk 3 above)
Comment (BeritJanssen) [reply backend/visualization/term_frequency.py]: So it is!

**Pull request merged**